Introduce PanacheQuery#select() method

mkouba opened the pull request:
The main motivation was to be able to reuse Panache queries when fetching data for a GraphQL query, i.e. something like:
https://github.com/mkouba/quarkus-quickstarts/blob/panache-graphql/hibernate-orm-panache-quickstart/src/main/java/org/acme/hibernate/orm/panache/FruitGraphqlResource.java#L24-L28

mkouba committed:
Add PanacheQuery#project() that accepts row mapper

- the function is used to create a DTO from each result line

Co-authored-by: Justin Lee <evanchooly@users.noreply.github.com>

loicmathieu commented:
We already have `project(Class)` for this use case.

geoand commented:
I like it, but this is really up to @FroMage and @loicmathieu

loicmathieu commented:
I'm not against it but if we accept it we may think about deprecating the `project(Class)` method.
If it's for a GraphQL specific use case, we can also provides this as a specialization for GraphQL and avoid adding it on the public API (like the specialization for Spring Data).

Let's see what @FroMage think about it.

mkouba commented:
Fair enough. I don't think we should depreacate the `project(Class)` as it's a good fit for a different use case. Also I don't think it's a GraphQL-specific use case, although it's a good example ;-).

mkouba commented:
@Sanne @gsmet WDYT?

FroMage commented:
> Not really. That's a different use case. First of all, for project() a "projection class" is needed for every combination of fields. 

It is different, but not very different. ATM you're right we need a different class per set of fields, but we could solve this with allowing to use alternate constructors.

> Furthermore, it's implemented as a "dynamic instantiation". In my use case, I don't have a DTO projection class but a dynamic set of fields and the result transformer logic is more powerful.

It is different, but:

1/ JPA already has support for that, as https://www.baeldung.com/hibernate-query-to-custom-class#resulttransformer tells me, but it's deprecated. It'd be nice to know why it's deprecated, perhaps @Sanne can find out?
2/ I'd rather make this per-line than take a list, it feels more proper
3/ This sort of API means the lambdas will have `List<Object>` as input which is horrible. I know it's due to the limitations of the current Panache hacks, but it's still super frustrating, we should at least try to find a way around this.

And finally, I think we should get back to the original use-case that prompted this feature and discuss graphql, because the example you show in https://github.com/mkouba/quarkus-quickstarts/blob/panache-graphql/hibernate-orm-panache-quickstart/src/main/java/org/acme/hibernate/orm/panache/FruitGraphqlResource.java#L24-L28 tells me you will transform managed entities with all fields set into non-managed entities with only specific fields set (that the user required), to be used as DTOs.

That tells me:

1/ we're loading stuff from the DB we already know we don't need, so that's wasteful, we should be able to do better
2/ we should really organise a discussion around facilitating DTOs, because @Sanne @geoand and @stuartwdouglas also have ideas about how to do/not do that. and perhaps put this PR on hold until then?

FroMage reviewed (changes requested):
I've actually left my review as a comment.

FroMage commented:
And BTW, big +1 on trying to improve the graphql experience @mkouba!

mkouba commented:
> It is different, but not very different. ATM you're right we need a different class per set of fields, but we could solve this with allowing to use alternate constructors.

@FroMage Hm, this approach does not scale well... also the set of fields may not be known at compile time.

> 2/ I'd rather make this per-line than take a list, it feels more proper

Done.

> 3/ This sort of API means the lambdas will have List<Object> as input which is horrible.

Well, it's a low-level API and since `javax.persistence.Query.getResultList()` returns `List` I think it's more or less OK ;-).

> 1/ we're loading stuff from the DB we already know we don't need, so that's wasteful, we should be able to do better

Yes, that's what this PR is all about. Select the fields we're interested in.

2/ we should really organise a discussion around facilitating DTOs

It's not only about DTOs, take a look at this test: https://github.com/quarkusio/quarkus/pull/12708/files#diff-25fe5182729f25a478d9a74d0c7bd4413b43c4fdbd383b4d037c479b0c0ab095R50

FroMage commented on extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheQuery.java:
```diff
@@ -40,6 +41,19 @@
+     * @param resultMapper
+     * @return a new query with the same state as the previous one (params, page, range, lockMode, hints, ...).
+     */
+    public <T> PanacheQuery<T> select(List<String> fields, Function<Object, T> resultMapper);
```
If we change the order of the parameters, we can get a more type-safe API:

```java
public <T> PanacheQuery<T> select(Function<Object, T> rowMapper, String field);
public <T> PanacheQuery<T> select(Function<Object[], T> rowMapper, String... fields);
```

It's a bit backwards, but it saves having to cast to an array when using more than one field. WDYT?

mkouba replied:
I've decided to only add  `public <T> PanacheQuery<T> project(Function<Object[], T> rowMapper, String... properties)` so that a result row is always represented by `Object[]`. I think that it's a more clear and "unified" approach.

FroMage commented on extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheQuery.java:
```diff
@@ -40,6 +41,19 @@
+     * 
+     * @param <T>
+     * @param fields
+     * @param resultMapper
```
Can we call this a `rowMapper` this way it's clear that this is called for every row. I find "result" to be a little bit ambiguous about whether this is called per-row or once for all rows.

FroMage commented on extensions/panache/hibernate-orm-panache/deployment/src/test/java/io/quarkus/hibernate/orm/panache/deployment/test/select/QuerySelectTest.java:
```diff
@@ -0,0 +1,71 @@
+        assertEquals("Jo", query.firstResult().name);
+
+        List<Integer> heights = Person.find("order by height desc").select(singletonList("height"), result -> (Integer) result)
+                .list();
```
So this should just work OOTB without `select` if we drop the type argument constraint:

```java
List<Integer> heights = Person.find("height order by height desc").list();
```

We should drop that type argument constraint.

mkouba replied:
Sounds good. What class should I look at?

FroMage replied:
I think it's the `PanacheQuery` class, or methods. And `list`/`stream`/`find` methods. Note that I haven't tested this, we have to check that the tests still compile if we drop it.

mkouba replied:
Ok, we can fix this in a followup PR.

FroMage reviewed (changes requested):
OK, so it's a row mapper. You've changed my mind, I think this is useful and a valid alternative to constructors.

I've added a question about the API, but this will be a good addition.

I'd leave `project` for now as they're not exactly the same, and it's not clear exactly which API is a better fit for which use-case.

loicmathieu commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -103,7 +110,30 @@ private CommonPanacheQueryImpl(CommonPanacheQueryImpl<?> previousQuery, String n
+        if (resultMapper == null) {
+            throw new PanacheQueryException("A result mapper must be set");
+        }
+        StringBuilder select = new StringBuilder("select ");
```
A query can start by `select` so you cannot blindly add a select clause in front of it.
I think this case should be tested and leads to a `PanacheQueryException`.

mkouba replied:
Hm, `CommonPanacheQueryImpl.project(Class<T>)` does the same. So how should we correctly detect a problematic query?

loicmathieu replied:
I think both methods should throw an exception in this case.

mkouba replied:
I've added a simple check that throws an exception if an existing query starts with `select`. I forgot to add a test though ;-). I'll add it in a minute.

loicmathieu commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -103,7 +110,30 @@ private CommonPanacheQueryImpl(CommonPanacheQueryImpl<?> previousQuery, String n
+        return new CommonPanacheQueryImpl<>(this, select.toString() + query, "select count(*) " + query, null);
+    }
+
+    public <T> CommonPanacheQueryImpl<T> select(List<String> columns, Function<Object, T> resultMapper) {
```
Shouldn't it be `select(List<String> columns, Function<Entity, T> resultMapper)` ?

mkouba replied:
I don't think so. We don't have to return a list of `Entity`... it could be anything, based on the mapper function.

loicmathieu replied:
OK, I just understood it, it maps from an Object (the row from the resultset) to a new return type `T`.

loicmathieu commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -227,7 +257,15 @@ private String countQuery(String selectQuery) {
         Query jpaQuery = createQuery();
         try (NonThrowingCloseable c = applyFilters()) {
-            return jpaQuery.getResultList();
+            List<T> results = jpaQuery.getResultList();
+            if (resultMapper != null) {
+                List<T> mappedResults = new ArrayList<>(results.size());
+                for (Object result : results) {
+                    mappedResults.add((T) resultMapper.apply(result));
+                }
+                results = mappedResults;
+            }
+            return results;
```
Something seems wrong here are you map a type `T` to another type `T`.
It works due to the fact that generics are erased at runtime but it looks like someting is not correct. 
Maybe it's linked to the other remarks I noticed about `Object` being used in the signature of the resultMapper.

`List<T> results = jpaQuery.getResultList();` is not correct as `getResultList();` returns a raw List, so either use `List<Object>` or `List<?>`.

loicmathieu replied:
So, I think a more correct code and understandable code should look like:

```
if (resultMapper == null) return jpaQuery.getResultList();
List results = jpaQuery.getResultList();
List<T> mappedResults = new ArrayList<>(results.size());
for (Object result : results) {
    mappedResults.add((T) resultMapper.apply(result));
}
return mappedResults;
```

Is the cast to `T` needed ? The resutMapper should have the correct return type if generics are correctly handled right ?

mkouba replied:
Yes, the `List<T>` should be just `List` because we map `Object` to `T`. Anyway, I'm just in the middle of renaming and refactoring. So let's discuss the final code...

mkouba replied:
BTW the `List<T> results = jpaQuery.getResultList()` is used all around this class...

mkouba replied:
Generics hell...

mkouba replied:
> Is the cast to T needed ?

It is for some reason. At least in my Eclipse IDE.

loicmathieu reviewed (changes requested):
`project` and `select` do the same: they project the results on a restricted list of column. One by using constructor expression, the other by using a row mapper.

The JavaDoc needs to reflect that they do the same with different means.

I strongly suggest to both name them `project` so to rename `select` and add an overload of `project` so it's clearer that they are both doing the same

mkouba commented:
> `project` and `select` do the same: they project the results on a restricted list of column. One by using constructor expression, the other by using a row mapper.
> 
> The JavaDoc needs to reflect that they do the same with different means.
> 
> I strongly suggest to both name them `project` so to rename `select` and add an overload of `project` so it's clearer that they are both doing the same

I don't have a strong opinion. The new method does not use dynamic instantiation and a projection class under the hood but it's true that it does almost the same as `project()` so I'm +0.

mkouba commented:
The method signature is now: `project(Function<Object[], T> rowMapper, String... properties)`

loicmathieu commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -79,15 +87,17 @@ private CommonPanacheQueryImpl(CommonPanacheQueryImpl<?> previousQuery, String n
         }
-
+        if (query != null && query.toLowerCase().startsWith(SELECT)) {
+            throw new PanacheQueryException("Cannot perform projection on an existing SELECT query: " + query);
```
```suggestion
            throw new PanacheQueryException("Cannot perform projection on a query with an existing select clause: " + query);
```

loicmathieu replied:
Same for the other occurence

mkouba commented:
The JVM Tests - JDK 8 failure does not seem to be related:
```
2020-11-24T13:03:43.2964548Z [ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 14.861 s <<< FAILURE! - in io.quarkus.vertx.http.shutdown.ShutdownTest
2020-11-24T13:03:43.2971254Z [ERROR] io.quarkus.vertx.http.shutdown.ShutdownTest  Time elapsed: 14.861 s  <<< FAILURE!
2020-11-24T13:03:43.2976451Z org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
```

mkouba commented:
@loicmathieu @FroMage Guys, any objections? This PR is getting old ;-).

FroMage reviewed (approved):
LGTM. @loicmathieu do we need an equivalent in Mongo Panache? @evanchooly same question for Kotlin?

FroMage commented:
It _is_ getting old. Sorry about that.

loicmathieu commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -256,19 +295,23 @@ private String countQuery(String selectQuery) {
         Query jpaQuery = createQuery(2);
         try (NonThrowingCloseable c = applyFilters()) {
-            List<T> list = jpaQuery.getResultList();
+            List<T> list = mapRows(jpaQuery.getResultList());
```
Again, you map the full result list and git the first, this is suboptimal

loicmathieu commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -79,15 +87,17 @@ private CommonPanacheQueryImpl(CommonPanacheQueryImpl<?> previousQuery, String n
             throw new PanacheQueryException("Unable to perform a projection on a named query");
         }
-
+        if (query != null && query.toLowerCase().startsWith(SELECT)) {
+            throw new PanacheQueryException("Cannot perform projection on a query with an existing select clause: " + query);
+        }
```
Can you add the same check on the other projet method ?

mkouba replied:
It's actually already there ;-)

loicmathieu replied:
OK, cool :)

loicmathieu commented on extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheQuery.java:
```diff
@@ -40,6 +41,20 @@
+     * @param rowMapper Maps a result row to a DTO object.
+     * @param properties
+     * @return a new query with the same state as the previous one (params, page, range, lockMode, hints, ...).
+     * @see #project(Class)
```
Can you also add a `@see` on the doc of the other projet method for a better discoverability ?

loicmathieu commented:
@mkouba sorry for the late review, I add another round of remarks.

As you said, the same must be done for kotlin.

For MongoDB it's not mandatory but we try to keep both API in sync if it seems legi.
 MongoDB projection is a different story as we use `TDOCUMENT` collection (typed collection) and not `Document` collection (raw collection using the bson Document type to express a result item) so a type is bound to a collection and we cannot get a raw list of items from it easily.
If you feel that something should be done on this area, you can create an issue to discuss it.

mkouba commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -239,12 +275,15 @@ private String countQuery(String selectQuery) {
+            if (results.isEmpty()) {
+                return null;
+            }
+            return mapRows(results).get(0);
```
That's a very good point!

loicmathieu commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -256,7 +298,12 @@ private String countQuery(String selectQuery) {
         try (NonThrowingCloseable c = applyFilters()) {
-            return (T) jpaQuery.getSingleResult();
+            T result = (T) jpaQuery.getSingleResult();
+            if (rowMapper != null) {
+                Object[] row = result.getClass().isArray() ? (Object[]) result : new Object[] { result };
+                result = (T) rowMapper.apply(row);
+            }
+            return result;
```
You should be able to use the new `mapRow` private function right ?

loicmathieu reviewed (approved):
LGTM except one small code reuse opportunity.

@evanchooly if you have some time can you check the new method signature added to the Kotling PanacheQuery version ?

evanchooly commented on extensions/panache/hibernate-orm-panache-kotlin/runtime/src/main/kotlin/io/quarkus/hibernate/orm/panache/kotlin/PanacheQuery.kt:
```diff
@@ -27,6 +28,13 @@ interface PanacheQuery<Entity: Any> {
+     *
+     * @return a new query with the same state as the previous one (params, page, range, lockMode, hints, ...).
+     */
+    fun <NewEntity: Any> project(rowMapper: Function<Array<Object>, NewEntity>, vararg properties: String): PanacheQuery<NewEntity>
```
For the kotlin side, i would reverse the argument order.  kotlin doesn't require that `vararg` arguments go last and reversing the order would allow for more idiomatic kotlin like this:

```kotlin
val query = Person.find("order by height desc") 
                .project("name", "surname") { result -> 
                    User user = new User();
                    user.name = result[0] as String;
                    user.surname = result[1] as String;
                    return user;
                }
```

mkouba replied:
Hm, I can change that but I thought that we should keep the kotlin API as close as possible.

evanchooly replied:
I try to do the same but sometimes it's "better" to make things idiomatic, imo.

loicmathieu replied:
The main purpose of having a different interface for Kotlin and Java is to allow idiomatic Kotlin API so I'm +1 on this one.

evanchooly commented on extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheQuery.java:
```diff
@@ -37,9 +38,24 @@
+     * @return a new query with the same state as the previous one (params, page, range, lockMode, hints, ...).
+     * @see #project(Class)
+     */
+    public <T> PanacheQuery<T> project(Function<Object[], T> rowMapper, String... properties);
```
I would argue here, as well as in the kotlin form, to reverse the arguments.  Take a `List<String>` instead so devs can put the more complicated lambda at the end.  Or better yet on a new lilne.  It would be far more readable that way.  As it is now, the field names tend to get lost in the lambda noise.

```java
        List<Person> namesAndSurnames = Person.find("order by height desc")
                .project(Arrays.asList("name", "surname"), 
                    result -> {
                        Person p = new Person();
                        p.name = result[0].toString();
                        p.surname = result[1].toString();
                        return p;
                    }).stream().collect(Collectors.toList());
```

mkouba replied:
Well, that was my original proposal but others thought that we should use varargs... @FroMage @loicmathieu WDYT?

evanchooly replied:
My problem with the varargs (beside the ordering) is that it makes it easy to have an empty set with out noticing.  There must be at least one argument which can't be enforced with varargs.  Yes, one could pass `List.of()` and get the same empty set but that's a much more conscious choice and less prone to accidentally fat fingering the invocation.

But the readability is the ultimate big win for me.

loicmathieu replied:
I'll let @FroMage give his advise on this one, he's the one that ask to reverse the parameters and use varags but @evanchooly made a point, if the mapper is complex (and it will be if people map to a POJO) better have it at the end for readibility. So this seems to be a compromise of redeability vs writability ;)

FroMage replied:
I prefer varargs because that _is_ the idiomatic way to do it in Java, even though I agree the order is inconvenient. Having to use `Arrays.asList` for _every_ call is a code smell: it means we're not using the language right.

As for requiring at least one argument, the idiomatic (albeit lame) way to do this in Java is:

```java
void foo(String one);
void foo(String one, String... more);
```

I do agree that we can reverse the order in Kotlin if that's idiomatic.

mkouba commented on extensions/panache/hibernate-orm-panache-common/runtime/src/main/java/io/quarkus/hibernate/orm/panache/common/runtime/CommonPanacheQueryImpl.java:
```diff
@@ -223,28 +258,35 @@ private String countQuery(String selectQuery) {
         Query jpaQuery = createQuery();
         try (NonThrowingCloseable c = applyFilters()) {
-            return jpaQuery.getResultList();
+            List<T> results = jpaQuery.getResultList();
```
The `rowMapper` is checked in the `mapRows()` method.

evanchooly replied:
right.  there's just a lot of inconsistency across the various invocations.

mkouba replied:
Well, I wouldn't call it inconsistency. We do try to optimize the code a little bit. For `stream()` it wouldn't make sense to map the result if no `rowMapper` is set, for other `mapRow()` usages it's easier to check the `rowMapper` inside the method.

mkouba commented:
I've asked @evanchooly to update this PR with the kotlin-related changes...

evanchooly commented:
My dev env was a bit of a mess so it took me a minute to clean it up.  but i see a problem with rearranging.  The interface is in Kotlin but the impl is in Java.  So those varargs have to go last to satisfy Java.  I still think we should a `List` there instead.  varargs imply that passing 0 is ok and it's not in this case.  And a `List` makes the API more workable, IMO, with the ability to move that `Function` to the end which is a big gain even in Java.

mkouba commented:
>  And a List makes the API more workable, IMO, with the ability to move that Function to the end which is a big gain even in Java.

I tend to agree. In JDK9+ you can use the `List.of()` methods that makes it even easier.  And the varargs is more error prone because you can forget to add the properties.

@FroMage I know that you're pushing the varargs as Java "idiomatic approach" but don't you think these arguments are valid?

FroMage commented:
When I doubt myself, I ask @gavinking for his opinion :)

mkouba commented:
I've rebased this PR and changed the signature to `io.quarkus.hibernate.orm.panache.PanacheQuery.project(List<String>, Function<Object[], T>)`.  I believe that the current version reads better, _"find persons, select name and surname and apply the row transformer"_ VS _"find persons, apply row transformer but select only name and surname"_.

In any case, this is a small "feature" and the difference is negligible... so unless there are some strong objections I'd like to have this PR merged at last ;-).

The current API:
```java
 Person.find("order by height desc")
                .project(List.of("name", "surname"), result -> {
                    Person p = new Person();
                    p.name = result[0].toString();
                    p.surname = result[1].toString();
                    return p;
                });
```
previous version:
```java
         Person.find("order by height desc")
                .project(result -> {
                    Person p = new Person();
                    p.name = result[0].toString();
                    p.surname = result[1].toString();
                    return p;
                }, "name", "surname");
```

gavinking commented:
Well this just looks like the worlds worst way to write:

```java
createQuery("select new Person(name, surname) from Person order by height desc").getSingleResult()
```

I thought the purpose of this work was to make it *easier* to write queries, not *harder*.

Anyway, it just looks awful to me.

mkouba commented:
> Well this just looks like the world's worst way to write:
> 
> ```java
> createQuery("select new Person(name, surname) from Person order by height desc").getSingleResult()
> ```
> 
> I thought the purpose of this work was to make it _easier_ to write queries, not _harder_.
> 
> Anyway, it just looks awful to me.

Well, that was just an example from the test. You don't need to use the `Person` class at all. You can transform the row to any object. Also `select new Person(name, surname)` is nice but requires a specific constructor for each combination of properties if I'm not mistaken.

gavinking commented:
> You can transform the row to any object. 

Sure, but you can do that with `select new` too. (In fact that's the more common case.)

> but requires a specific constructor for each combination of properties if I'm not mistaken.

Yes, that's true. But in practice the things you're sticking the data in are essentially *always* classes you control, so it's essentially never a problem to add a constructor.

Further, it's unclear to me why it's much of an advantage to map tuples to classes as you're building the result list, rather than just using `Stream.map()` on the list you get back from `getResultList()`. I mean, sure, I can imagine up some crazy scenarios in which it might make some sense, but I don't think we're supposed to be addressing crazy scenarios at this stage.

All, that said, it would certainly be nice if you would have the alternative of using a static method for this from within HQL. (That's an enhancement we could pretty easily make in Hibernate.)

gavinking commented:
```java
createQuery("select name, surname from Person order by height desc").getSingleResult().map(result -> {
                    Person p = new Person();
                    p.name = result[0].toString();
                    p.surname = result[1].toString();
                    return p;
                })
```

Or something like that. (I didn't try to compile it.)

gavinking commented:
> it would certainly be nice if you would have the alternative of using a static method for this from within HQL. 

i.e. it would be nice to be able to write

```java
createQuery("select PersonUtil.create(name, surname) from Person order by height desc") .getResultList()
```

mkouba commented:
Ok, so the original idea was to _extend the Panache query API_ because currently you can't do `Person.find("select name from Person order by name asc").page(...).stream()...` etc. 

You will get the obvious `ClassCastException`. `class java.lang.String cannot be cast to class io.quarkus.hibernate.orm.panache.PanacheEntityBase` for single column and  `class [Ljava.lang.Object; cannot be cast to class io.quarkus.hibernate.orm.panache.PanacheEntityBase` for multiple columns.

With this change you can do e.g. something like:
```java
PanacheQuery<Map<String, Object>> paged = Person.find("order by name asc")
                .project(List.of("surname", "height"), result -> {
                    // TOO turn Object[] into a map...                    
                    return map;
                }).page(Page.ofSize(10));

paged.list().stream().forEach(System.out::println); // ->  first page
paged.nextPage().stream().forEach(System.out::println); // -> next page
```
I don't think that this is that crazy but if you guys have a better idea I'm all ears.

gavinking commented:
Well I guess it's "crazy" in the sense that you're inventing complicated and verbose APIs to write the equivalent of:

    select new map(surname, height) from Person order by name asc

And you're guiding users down the path of thinking they have to avoid just writing their queries in HQL, which is always going to be the easiest thing to do in the case of any query that's not completely trivial.

You're making them invest their efforts in learning an API that is always going to be fundamentally crippled compared to a real query language, instead of getting comfortable with the query language itself.

So the good solution is to make it easy for users to drop back into plain HQL whenever they're using Panache. I had understood this to be already possible.

gavinking commented:
So here's the thing. I've written 3 of these sort of query APIs in the past, and I've recently toyed with a fourth. And _every single time_ what I ultimately realized is that except for the _most trivial queries_, they're always much worse than HQL. And at this point I'm ashamed of every single one of them! 

On the other hand, I only had to design HQL once. Sure, we're currently working on the third *reimplementation* of HQL, and fixing some things which suck about it, but it's the same basic language. A user who learned HQL *20 years ago* in Hibernate 1.0 could come to Quarkus now and know *exactly what to do*. Any other persistence or query API they had learned back then is now completely and utterly dead. 

Is that an accident? I don't think so. I think the simple truth of the matter is that for all the disadvantages of embedded languages (and for sure they're there), as soon as you start writing a nontrivial query, a query language is what you need. The APIs all suck. The very best query API is jOOQ and while on the one hand I kinda love what he's done with it, the simple fact is that unless I had very specific requirements for cross-database portability, I quite honestly would *never* use that in preference to just writing SQL in SQL!

Now, all that said, if you insist on providing an API, I guess there's one decent alternative that might work and not be a quagmire. If you could provide an easy way to obtain and use bits of the JPA criteria query API from within the panache object, that might work well. Currently using the criteria API is sucky because you have to write shit like `criteriaBuilder.equal(itemRoot.get(Person_.surname), "King")`. But if Panache could expose a field like `_surname` or perhaps a method like `_surnameEquals()` from within the `Person` class, then that that would make it *way* more usable. And then you would just make it easy to consume those objects from `find()`.

The advantage of that approach is that the criteria API, for all its complexity, is an actual complete model of JPQL and has _everything_ there.

mkouba commented:
> ...
> 
> Is that an accident? I don't think so. I think the simple truth of the matter is that for all the disadvantages of embedded languages (and for sure they're there), as soon as you start writing a nontrivial query, a query language is what you need. The APIs all suck. The very best query API is jOOQ and while on the one hand I kinda love what he's done with it, the simple fact is that unless I had very specific requirements for cross-database portability, I quite honestly would _never_ use that in preference to just writing SQL in SQL!

I used to use Hibernate Criteria to build dynamic queries and it was very practical compared to hand-written builders that just glue chunks of HQL together.

> 
> Now, all that said, if you insist on providing an API, I guess there's one decent alternative that might work and not be a quagmire. If you could provide an easy way to obtain and use bits of the JPA criteria query API from within the panache object, that might work well. 

I don't insist on providing an API, I just thought that it would be a small and useful addition to panache query API. I'll let @FroMage to decide and possibly close this PR. I'm not an expert in this area.

> ...

FroMage commented:
> Ok, so the original idea was to extend the Panache query API because currently you can't do Person.find("select name from Person order by name asc").page(...).stream()... etc.

> You will get the obvious ClassCastException. class java.lang.String cannot be cast to class io.quarkus.hibernate.orm.panache.PanacheEntityBase for single column and class [Ljava.lang.Object; cannot be cast to class io.quarkus.hibernate.orm.panache.PanacheEntityBase for multiple columns.

This is only due to the type argument constraint, and we should drop it.

FroMage commented:
I guess with this API you can do stuff you can't in HQL:

```java
int counter = 0;
PanacheQuery<SomeObject> paged = Person.find("order by name asc")
                .project(List.of("surname", "height"), result -> {
String name = String.format("%020s", (String)result[0])
int height = height * someJavaLogic();
                    return SomeObject.specialApi().makeItSo( name, height, counter++);
                }).page(Page.ofSize(10));

paged.list().stream().forEach(System.out::println); // ->  first page
paged.nextPage().stream().forEach(System.out::println); // -> next page

```

If you add support in HQL for static functions, we might be able to get close to this by putting the mapper function in a static function, but you still lose all captured variables, and HQL is not turing complete, is it? (I hope it's not).

If you put this mapper on each streams you obtain from the query, then you have to reapply the map operation at every place you use the query (`paged.list().stream()` and `paged.nextPage().stream()` here).

So it's really like `map`, but for a query, applied to every list/stream that comes out of it.

I really think there's value in something like this, just as there's value in improving HQL's projection capabilities with static functions.

FroMage commented:
> But if Panache could expose a field like _surname or perhaps a method like _surnameEquals() from within the Person class, then that that would make it way more usable. And then you would just make it easy to consume those objects from find().

No, Panache is not APT: it can't add signatures to the object it instruments. Even APT can't, not without hacks.

evanchooly commented:
Is this still valid/wanted?  Should it be merged or closed?  I just found it again.  I'm surprised it's still here and still mergeable.  :)

mkouba commented:
I'm going to close this PR. No one seemed inclined to "accept" this API change...

PS. @evanchooly I also wonder it's still mergeable :-)
